File: 49. Group Anagrams/49.cpp
// L = maximum length of string
// time: O(n*Llog(L)) space: O(n*L) stored in map
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // categorize by sorted string
        unordered_map<string, vector<string>> anagram_map;
        for(string str: strs){
            string sorted = str;
            sort(sorted.begin(), sorted.end());
            anagram_map[sorted].push_back(str);
        }

        vector<vector<string>> groups;
        for(auto it: anagram_map){
            groups.push_back(it.second);
        }

        return groups;
    }
};

File: 49. Group Anagrams/49.py
"""
time: O(n * k log k)
space: O(nk)
"""
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagram_dic = {}
        
        for word in strs:
            key = "".join(sorted(word))
            
            if key in anagram_dic:
                anagram_dic[key].append(word)
            else:
                anagram_dic[key] = [word]
        
        ans = []
        for key, value in anagram_dic.items():
            ans.append(value)
        return ans

"""
anagram_dic = {sorted_unique_anagram: [anagram strings]}
k: maximum lenght of str in strs -> sort O(k log k)
n: elemets in strs
"""
